File: CSharp/ThrustRTCSharp/FakeVectors.cs
using System;
using ThrustRTCLR;
using System.Runtime.InteropServices;

namespace ThrustRTCSharp
{
    public class DVConstant : DVVectorLike
    {
        public DVConstant(DeviceViewable dvobj, long size = -1)
            : base(Native.dvconstant_create(dvobj.m_cptr, (ulong)size))
        {
            m_dvobj = dvobj;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_dvobj = null;
            base.Dispose(disposing);
        }

        private DeviceViewable m_dvobj;
    }

    public class DVCounter : DVVectorLike
    {
        public DVCounter(DeviceViewable dvobj_init, long size = -1)
            : base(Native.dvcounter_create(dvobj_init.m_cptr, (ulong)size))
        {
            m_dvobj_init = dvobj_init;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_dvobj_init = null;
            base.Dispose(disposing);
        }

        private DeviceViewable m_dvobj_init;
    }

    public class DVDiscard : DVVectorLike
    {
        public DVDiscard(string elem_cls, long size = -1)
            : base(Native.dvdiscard_create(Marshal.StringToHGlobalAnsi(elem_cls), (ulong)size))  {}
    }

    public class DVPermutation : DVVectorLike
    {
        public DVPermutation(DVVectorLike vec_value, DVVectorLike vec_index)
            : base(Native.dvpermutation_create(vec_value.m_cptr, vec_index.m_cptr))
        {
            m_vec_value = vec_value;
            m_vec_index = vec_index;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_vec_value = null;
            m_vec_index = null;
            base.Dispose(disposing);
        }

        private DVVectorLike m_vec_value;
        private DVVectorLike m_vec_index;
    }

    public class DVReverse : DVVectorLike
    {
        public DVReverse(DVVectorLike vec_value)
            : base(Native.dvreverse_create(vec_value.m_cptr))
        {
            m_vec_value = vec_value;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_vec_value = null;
            base.Dispose(disposing);
        }

        private DVVectorLike m_vec_value;
    }

    public class DVTransform : DVVectorLike
    {
        public DVTransform(DVVectorLike vec_in, string elem_cls, Functor op)
            : base(Native.dvtransform_create(vec_in.m_cptr, Marshal.StringToHGlobalAnsi(elem_cls), op.m_cptr) )
        {
            m_vec_in = vec_in;
            m_op = op;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_vec_in = null;
            m_op = null;
            base.Dispose(disposing);
        }

        private DVVectorLike m_vec_in;
        private Functor m_op;
    }

    public class DVZipped : DVVectorLike
    {
        static IntPtr create(DVVectorLike[] vecs, string[] elem_names)
        {
            IntPtr[] p_vecs = new IntPtr[vecs.Length];
            for (int i = 0; i < vecs.Length; i++)
                p_vecs[i] = vecs[i].m_cptr;
            IntPtr[] p_elem_names = new IntPtr[elem_names.Length];
            for (int i = 0; i < elem_names.Length; i++)
                p_elem_names[i] = Marshal.StringToHGlobalAnsi(elem_names[i]);

            return Native.dvzipped_create(p_vecs, p_elem_names);
        }

        public DVZipped(DVVectorLike[] vecs, string[] elem_names)
            : base (create(vecs, elem_names))
        {
            m_vecs = vecs;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_vecs = null;
            base.Dispose(disposing);
        }

        private DVVectorLike[] m_vecs;
    }

    public class DVCustomVector : DVVectorLike
    {
        static IntPtr create(DeviceViewable[] objs, string[] name_objs, string name_idx, string code_body, string elem_cls, long size, bool read_only)
        {
            CapturedDeviceViewable_clr[] arg_map = new CapturedDeviceViewable_clr[objs.Length];
            for (int i = 0; i < objs.Length; i++)
            {
                arg_map[i].obj = objs[i].m_cptr;
                arg_map[i].obj_name = Marshal.StringToHGlobalAnsi(name_objs[i]);
            }

            return Native.dvcustomvector_create(arg_map,
                Marshal.StringToHGlobalAnsi(name_idx),
                Marshal.StringToHGlobalAnsi(code_body),
                Marshal.StringToHGlobalAnsi(elem_cls), (ulong)size, read_only);
        }

        public DVCustomVector(DeviceViewable[] objs, string[] name_objs, string name_idx, string code_body, string elem_cls, long size = -1, bool read_only = true)
            : base(create(objs, name_objs, name_idx, code_body, elem_cls, size, read_only))
        {
            m_objs = objs;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposed) return;
            m_objs = null;
            base.Dispose(disposing);
        }

        private DeviceViewable[] m_objs;

    }
}

File: CSharp/ThrustRTCLR/FakeVectors.cpp
#include "stdafx.h"
#include "ThrustRTCLR.h"
#include "fake_vectors/DVConstant.h"
#include "fake_vectors/DVCounter.h"
#include "fake_vectors/DVDiscard.h"
#include "fake_vectors/DVPermutation.h"
#include "fake_vectors/DVReverse.h"
#include "fake_vectors/DVTransform.h"
#include "fake_vectors/DVZipped.h"
#include "fake_vectors/DVCustomVector.h"

namespace ThrustRTCLR
{
	template<typename T>
	inline T* just_cast_it(IntPtr p)
	{
		return (T*)(void*)p;
	}

	IntPtr Native::dvconstant_create(IntPtr p_dvobj, size_t size)
	{
		DeviceViewable* dvobj = just_cast_it<DeviceViewable>(p_dvobj);
		return (IntPtr)(new DVConstant(*dvobj, size));
	}

	IntPtr Native::dvcounter_create(IntPtr p_dvobj_init, size_t size)
	{
		DeviceViewable* dvobj_init = just_cast_it<DeviceViewable>(p_dvobj_init);
		return (IntPtr)(new DVCounter(*dvobj_init, size));
	}

	IntPtr Native::dvdiscard_create(IntPtr p_elem_cls, size_t size)
	{
		const char* elem_cls = just_cast_it<const char>(p_elem_cls);
		return (IntPtr)(new DVDiscard(elem_cls, size));
	}

	IntPtr Native::dvpermutation_create(IntPtr p_vec_value, IntPtr p_vec_index)
	{
		DVVectorLike* vec_value = just_cast_it<DVVectorLike>(p_vec_value);
		DVVectorLike* vec_index = just_cast_it<DVVectorLike>(p_vec_index);
		return (IntPtr)(new DVPermutation(*vec_value, *vec_index));
	}

	IntPtr Native::dvreverse_create(IntPtr p_vec_value)
	{
		DVVectorLike* vec_value = just_cast_it<DVVectorLike>(p_vec_value);
		return (IntPtr)(new DVReverse(*vec_value));
	}

	IntPtr Native::dvtransform_create(IntPtr p_vec_in, IntPtr p_elem_cls, IntPtr p_op)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		const char* elem_cls = just_cast_it<const char>(p_elem_cls);
		Functor* op = just_cast_it<Functor>(p_op);
		return (IntPtr)(new DVTransform(*vec_in, elem_cls, *op));
	}

	IntPtr Native::dvzipped_create(array<IntPtr>^ p_vecs, array<IntPtr>^ p_elem_names)
	{
		int num_vecs = p_vecs->Length;
		std::vector<DVVectorLike*> vecs(num_vecs);
		for (int i = 0; i < num_vecs; i++)
			vecs[i] = just_cast_it<DVVectorLike>(p_vecs[i]);

		int num_elems = p_elem_names->Length;
		if (num_elems != num_vecs) return IntPtr::Zero;
		std::vector<const char*> elem_names(num_elems);
		for (int i = 0; i < num_elems; i++)
			elem_names[i] = just_cast_it<const char>(p_elem_names[i]);

		return (IntPtr)(new DVZipped(vecs, elem_names));
	}

	IntPtr Native::dvcustomvector_create(array<CapturedDeviceViewable_clr>^ p_arg_map, IntPtr p_name_idx, IntPtr p_code_body, IntPtr p_elem_cls, size_t size, bool read_only)
	{
		int num_params = p_arg_map->Length;
		std::vector<CapturedDeviceViewable> arg_map(num_params);
		for (int i = 0; i < num_params; i++)
		{
			arg_map[i].obj_name = just_cast_it<const char>(p_arg_map[i].obj_name);
			arg_map[i].obj = just_cast_it<DeviceViewable>(p_arg_map[i].obj);
		}
		
		const char* name_idx = just_cast_it<const char>(p_name_idx);
		const char* code_body = just_cast_it<const char>(p_code_body);
		const char* elem_cls = just_cast_it<const char>(p_elem_cls);

		return (IntPtr)(new DVCustomVector(arg_map, name_idx, code_body, elem_cls, size, read_only));
	}

}